src/lib/io/sol-ipm.c

```diff
@@ -0,0 +1,370 @@
+
+#include <errno.h>
+#include <inttypes.h>
+#include <stdlib.h>
+
+#ifndef SOL_LOG_DOMAIN
+#define SOL_LOG_DOMAIN &_sol_ipm_log_domain
+#endif
+
+#include "sol-log-internal.h"
+
+#include "ipm.h"
+#include "ipm/ipm_quark_se.h"
+
+#include "sol-ipm.h"
+#include "sol-log.h"
+#include "sol-macros.h"
+#include "sol-mainloop-zephyr.h"
+#include "sol-types.h"
+#include "sol-util.h"
+#include "sol-vector.h"
+
+SOL_LOG_INTERNAL_DECLARE_STATIC(_sol_ipm_log_domain, "ipm");
+
+#define SOL_IPM_ID_FREE_REMOTE 0
+
+/* Declare channels we are going to use.
+ */
+#ifdef CONFIG_X86
+QUARK_SE_IPM_DEFINE(message_ipm_receiver, IPM_CHANNEL_ARC_TO_X86, QUARK_SE_IPM_INBOUND);
+QUARK_SE_IPM_DEFINE(message_ipm_sender, IPM_CHANNEL_X86_TO_ARC, QUARK_SE_IPM_OUTBOUND);
+#elif CONFIG_ARC
+QUARK_SE_IPM_DEFINE(message_ipm_receiver, IPM_CHANNEL_X86_TO_ARC, QUARK_SE_IPM_INBOUND);
+QUARK_SE_IPM_DEFINE(message_ipm_sender, IPM_CHANNEL_ARC_TO_X86, QUARK_SE_IPM_OUTBOUND);
+#endif
+
+#define AS_PTR(_ptr) *(uintptr_t *)_ptr
+
+int sol_ipm_init(void);
+void sol_ipm_shutdown(void);
+
+/***** IPM Blobs *****/
+
+static void free_ipm_blob(struct sol_blob *blob);
+
+struct sol_blob_ipm_received {
+    struct sol_blob base;
+    struct sol_blob *remote;
+};
+
+struct sol_blob_ipm_sent {
+    struct sol_blob base;
+    uint32_t id;
+};
+
+static const struct sol_blob_type SOL_BLOB_TYPE_IPM_RECEIVED = {
+    SOL_SET_API_VERSION(.api_version = SOL_BLOB_TYPE_API_VERSION, )
+    SOL_SET_API_VERSION(.sub_api = 0, )
+    .free = free_ipm_blob,
+};
+
+/*** IPM Stuff ***/
+
+struct sol_ipm_receiver_entry {
+    uint32_t id;
+    void (*receive_cb)(void *cb_data, uint32_t id, struct sol_blob *message);
+    const void *data;
+    struct sol_blob *last_remote_blob; /* Hackish... I store last blob received here, to avoid having yet another list to store them */
+};
+
+struct sol_ipm_consumed_handler {
+    uint32_t id;
+    void (*message_consumed_cb)(void *data, uint32_t id, struct sol_blob *message);
+    const void *data;
+};
+
+static struct sol_vector ipm_receivers = SOL_VECTOR_INIT(struct sol_ipm_receiver_entry);
+static struct sol_vector ipm_consumed_handlers = SOL_VECTOR_INIT(struct sol_ipm_consumed_handler);
+static struct device *ipm_sender, *ipm_receiver;
+static uint32_t ipm_max_id;
+static bool initialised;
+
+static void
+free_remote_blob(struct sol_blob *blob)
+{
+    int r;
+
+    r = ipm_send(ipm_sender, 1, SOL_IPM_ID_FREE_REMOTE, &blob,
+        sizeof(uintptr_t));
+    if (r != 0) {
+        /* Should I care if return was EBUSY?*/
+        SOL_WRN("Could not send consumed message to remote core: %d", r);
+    }
+}
+
+static void
+free_ipm_blob(struct sol_blob *blob)
+{
+    struct sol_blob_ipm_received *ipm_blob = (struct sol_blob_ipm_received *)blob;
+
+    free_remote_blob(ipm_blob->remote);
+
+    /* Free local shadow */
+    free(blob);
+}
+
+extern unsigned char *heap_base;
+extern unsigned int heap_sz, heap_max;
+
+static void
+ipm_receiver_process(void *data)
+{
+    int idx;
```

Comment:
vectors use `uint16_t`, you should use it here as well. not sure why it did no trigger a signed x unsigned comparison warning